### ISSUE TITLE
type :foreman_smartproxy incompatible with `puppet generate types`

### ISSUE THREAD

**Issue by xorpaul** [theforeman/puppet-foreman#512]
Hello there,

I'm not sure if you can even do something about this, but here it goes...

I tried implementing the Puppetlabs fix for the Puppetserver environment isolation which is executing `puppet generate types --environment <ENVIRONMENT>` that generated the following file for the `foreman_smartproxy` type: [foreman_smartproxy.txt](https://github.com/theforeman/puppet-foreman/files/657369/foreman_smartproxy.txt)

This file is then the reason the catalog can't be compiled anymore as the `:url` property/parameter does use the ruby internal `URI.regexp` which `puppet generate types` does resolve to the full regex:

```
Puppet::Resource::Param(Pattern[/
        ([a-zA-Z][\-+.a-zA-Z\d]*):                           (?# 1: scheme)
        (?:
           ((?:[\-_.!~*'()a-zA-Z\d;?:@&=+$,]|%[a-fA-F\d]{2})(?:[\-_.!~*'()a-zA-Z\d;\/?:@&=+$,\[\]]|%[a-fA-F\d]{2})*)                    (?# 2: opaque)
        |
           (?:(?:
             \/\/(?:
                 (?:(?:((?:[\-_.!~*'()a-zA-Z\d;:&=+$,]|%[a-fA-F\d]{2})*)@)?        (?# 3: userinfo)
                   (?:((?:(?:[a-zA-Z0-9\-.]|%\h\h)+|\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}|\[(?:(?:[a-fA-F\d]{1,4}:)*(?:[a-fA-F\d]{1,4}|\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})|(?:(?:[a-fA-F\d]{1,4}:)*[a-fA-F\d]{1,4})?::(?:(?:[a-fA-F\d]{1,4}:)*(?:[a-fA-F\d]{1,4}|\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}))?)\]))(?::(\d*))?))? (?# 4: host, 5: port)
               |
                 ((?:[\-_.!~*'()a-zA-Z\d$,;:@&=+]|%[a-fA-F\d]{2})+)                 (?# 6: registry)
               )
             |
             (?!\/\/))                           (?# XXX: '\/\/' is the mark for hostport)
             (\/(?:[\-_.!~*'()a-zA-Z\d:@&=+$,]|%[a-fA-F\d]{2})*(?:;(?:[\-_.!~*'()a-zA-Z\d:@&=+$,]|%[a-fA-F\d]{2})*)*(?:\/(?:[\-_.!~*'()a-zA-Z\d:@&=+$,]|%[a-fA-F\d]{2})*(?:;(?:[\-_.!~*'()a-zA-Z\d:@&=+$,]|%[a-fA-F\d]{2})*)*)*)?                    (?# 7: path)
           )(?:\?((?:[\-_.!~*'()a-zA-Z\d;\/?:@&=+$,\[\]]|%[a-fA-F\d]{2})*))?                 (?# 8: query)
        )
        (?:\#((?:[\-_.!~*'()a-zA-Z\d;\/?:@&=+$,\[\]]|%[a-fA-F\d]{2})*))?                  (?# 9: fragment)
      /], 'url')
  ],
```

The error message can be simply generate with the `puppet parser validate` of the generated file:
```
puppet parser validate /tmp/.resource_types/foreman_smartproxy.pp
Error: Could not parse for environment production: Syntax error at '/' at /tmp/.resource_types/foreman_smartproxy.pp:34:37
```
Can you do something with this information?
Did you know of the environment isolation bug or the `puppet generate types` command before this text? I sure didn't two days ago. :smiley: 

I also addressed this issue to the Puppetlabs devs and linked this issue there as well: https://tickets.puppetlabs.com/browse/SERVER-94?focusedCommentId=385536&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-385536

Thanks in advance!

**Comment by domcleal** [comment]
The type could use a simpler regexp that's compatible with the Puppet language rather than the full Ruby version.

**Comment by SuDT** [comment]
@xorpaul Hi all, has anyone found a fix for this yet? Thanks

**Comment by SuDT** [comment]
OK so haven't completed full setup yet but the following change allowed Puppet to apply the catalogue successfully:

`Puppet::Resource::Param(Pattern["([a-zA-Z][\-+.a-zA-Z\d]*)"], 'url')`

**Comment by ekohl** [comment]
@xorpaul did you ever file the JIRA (PUP) issue for the parser generation?

**Comment by SuDT** [comment]
@ekohl link is in the original post

**Comment by ekohl** [comment]
@SuDT yes, but Puppetlabs (both Henrik and Eric) asked to create a separate issue for the parse issue and I can't find one. I don't think it'll be on their radar unless it's created.

**Comment by xorpaul** [comment]
@ekohl and @SuDT I just opened the ticket on the suggested Jira queue: https://tickets.puppetlabs.com/browse/PUP-7306

**Comment by ekohl** [comment]
@xorpaul thanks!

**Comment by SuDT** [comment]
Can you please also refer them to [my comment above](#issuecomment-276656153) as this simplification helped me progress

**Comment by domcleal** [comment]
Thanks, the issue has now been fixed in Puppet for 4.10.0 via PUP-7306, and I've tested that the generated type can be loaded. I'll go ahead and close this.
